File: bffcompiler.rb
#!/usr/bin/env ruby

require 'socket'
# TODO: Add functions support

# A list of allowed charachters.
ALLOWED = [
            '.', # Standard out
            ',', # Standard in
            '[', # Loop start
            ']', # Loop end
            '<', # Move back a cell
            '>', # Move ahead a cell
            '+', # Add 1 to the current cell
            '-', # Sub 1 from the current cell
            '#', # Open/close file
            '%', # Write val of current cell to file
            '!', # Read a char from file and set it as the current cell
            '@', # Open/close socket to localhost (on port 53333)
            '&', # Read 1 char from the socket and set as current cell
            '*', # Write val of current cell to socket
            '~', # goto cell[value of current cell]
            '$', # toggle character mode
            '^'  # open a server socket
          ]

def execute(filename)
  begin
    f = File.read(filename)
  rescue
    puts 'No such file.'
    exit
  end
  evaluate(f)
end

def evaluate(code)
  # cleanup the code (Remove anything not in ALLOWED) and then build a bracemap.
  code = cleanup(code.chars)
  bracemap = buildbracemap(code)
  handler = nil
  cursor = 0
  sock = nil
  serverSock = nil

  cells = [0]
  codeptr = 0
  cellptr = 0
  code_length = code.length
  cmode = true

  while codeptr < code_length
    command = code[codeptr]

    # Iterate through all the commands in the code and evaluate
    case command
    when '>' then cellptr += 1
                  cells.push(0) if cellptr == cells.length
    when '<' then cellptr = cellptr <= 0 ? 0 : cellptr - 1
    when '+' then cells[cellptr] = cells[cellptr] < 255 ? cells[cellptr] + 1 : 0
    when '-' then cells[cellptr] = cells[cellptr] > 0 ? cells[cellptr] - 1 : 255
    when '[' then if (cells[cellptr]).zero? then codeptr = bracemap[codeptr] end
    when ']' then if cells[cellptr] != 0 then codeptr = bracemap[codeptr] end
    when ',' then cells[cellptr] = $stdin.getc.ord
    when '.' then
      if cmode  
        $stdout.write cells[cellptr].chr
      else
        $stdout.write cells[cellptr].to_s
      end
    when '#' then
      if handler.nil?
        name = ""
        while cells[cellptr] != 0 do
            name = name + cells[cellptr].chr
            cellptr += 1
        end
        
        handler = File.open(name, 'r+')
        cursor = 0
      else
        handler.close
        handler = nil
      end
    when '%'
      if !handler.nil?
        handler.syswrite(cells[cellptr].chr)
      else
        $stderr.write "At #{codeptr}: ERROR - NO FILE IS OPEN\n"
        exit
      end
    when '!'
      if !handler.nil?
        if c = handler.getc
          cells[cellptr] = c[0].ord
          cursor += 1
        else
          cells[cellptr] = 0
        end
      else
        $stderr.write "At #{codeptr}: ERROR - NO FILE IS OPEN\n"
        exit
      end
    when '@'
      if sock.nil?
        name = ""
        while cells[cellptr] != 0 do
            name = name + cells[cellptr].chr
            cellptr += 1
        end

        sock = TCPSocket.new(name, 53333)
      else
        sock.close
        sock = nil
      end
    when '^'
        if serverSock.nil?
          serverSock = TCPServer.new(53333)
          sock = serverSock.accept
        else
          serverSock.close
          serverSock = nil
        end
    when '*'
      if !sock.nil?
        sock.write(cells[cellptr].chr)
      else
        $stderr.write "At #{codeptr}: ERROR - NO SOCK IS OPEN\n"
      end
    when '&'
      if !sock.nil?
        cells[cellptr] = sock.getc.ord
      else
        $stderr.write "At #{codeptr}: ERROR - NO SOCK IS OPEN\n"
      end
    when '~'
      if cells[cellptr] < code_length
        codeptr = cells[cellptr] - 1
      else
        $stderr.write "At #{codeptr}: ERROR - CELL VALUE OUT OF RANGE\n"
      end
    when '$'
      if !cmode then cmode = true else cmode = false end
    end

    codeptr += 1
  end
end

def cleanup(code)
  code.select { |a| ALLOWED.include? a }
end

def buildbracemap(code)
  temp_bracestack = []
  bracemap = {}
  code.each_with_index do |command, position|
    if command == '['
      temp_bracestack.push(position)
    elsif command == ']'
      start = temp_bracestack.pop
      bracemap[start] = position
      bracemap[position] = start
    end
  end
  bracemap
end

def main
  if ARGV.length == 1
    execute(ARGV[0])
  else
    puts "Usage: #{File.basename($PROGRAM_NAME)} filename"
  end
end

main
exit

File: bffcompiler.lua
-- Lua BF++ Interpretter
-- Samuel Taylor
local socket = require('socket')

TOKENS = {
    ['.'] = true, -- Standard out
    [','] = true, -- Standard in
    ['['] = true, -- Loop start
    [']'] = true, -- Loop end
    ['<'] = true, -- Move back a cell
    ['>'] = true, -- Move ahead a cell
    ['+'] = true, -- Add 1 to the current cell
    ['-'] = true, -- Sub 1 from the current cell
    ['#'] = true, -- Open/close file
    ['%'] = true, -- Write val of current cell to file
    ['!'] = true, -- Read a char from file and set it as the current cell
    ['@'] = true, -- Open/close socket to localhost (on port 53333)
    ['&'] = true, -- Read 1 char from the socket and set as current cell
    ['*'] = true, -- Write val of current cell to socket
    ['~'] = true, -- goto cell[value of current cell]
    ['$'] = true, -- toggle character mode
    ['^'] = true  -- open a server socket
}

function run(filename)
    local file = io.open(filename, "r")

    if file == nil then
        print("\x1B[31mERROR: No such file\x1B[0m")
    else
        eval(file)
    end 
end

function only_tokens(file)
    local chr = 1
    local tcode = {}

    while chr ~= nil do
        chr = file:read(1)
        tcode[#tcode + 1] = (TOKENS[chr] ~= nil and chr or nil)
    end

    return tcode
end

function bracemapper(code)
    local stack = {}
    local bracemap = {}

    for pos, tok in ipairs(code) do
        if tok == '[' then
            table.insert(stack, pos)
        elseif tok == ']' then
            local start = table.remove(stack)
            bracemap[start] = pos
            bracemap[pos] = start
        end
    end

    return bracemap
end

function eval(file)
    local code = only_tokens(file)
    local bracemap = bracemapper(code)

    local file = nil
    local sock = nil

    local tape = {0}

    tape.__index = function(table, key)
        table[key] = 0
        return 0
    end

    setmetatable(tape, tape)

    local codeptr = 1
    local tapeptr = 1

    local codelen = #code
    
    local switch = {
        ['>'] = function() 
            tapeptr = tapeptr + 1
            if tape[tapeptr] == nil then 
                tape[tapeptr] = 0 
            end
        end,
        ['<'] = function()
            tapeptr = tapeptr > 1 and tapeptr - 1 or 1
        end,
        ['+'] = function()
            tape[tapeptr] = tape[tapeptr] + 1
            if tape[tapeptr] > 255 then
                tape[tapeptr] = 0
            end
        end,
        ['-'] = function()
            tape[tapeptr] = tape[tapeptr] - 1
            if tape[tapeptr] < 0 then
                tape[tapeptr] = 255
            end
        end,
        ['['] = function()
            if tape[tapeptr] == 0 then
                codeptr = bracemap[codeptr]
            end
        end,
        [']'] = function()
            if tape[tapeptr] ~= 0 then
                codeptr = bracemap[codeptr]
            end
        end,
        [','] = function()
            tape[tapeptr] = string.byte(io.read(1))
        end,
        ['.'] = function()
            io.write(
                string.char(tape[tapeptr])
            )
        end,
        ['#'] = function()
            if file == nil then
                local name = {}
                while tape[tapeptr] ~= 0 do
                    table.insert(name, string.char(tape[tapeptr]))
                    tapeptr = tapeptr + 1
                end

                local str = table.concat(name)
                file = io.open(str)
            else
                io.close(file)
                file = nil
            end
        end,
        ['%'] = function()
            if file ~= nil then
                file:write(tape[tapeptr])
            else    
                error(string.format( "\x1B[31mERROR @ %d: NO FILE OPEN\x1B[0m\n", codeptr))
                return
            end
        end,
        ['!'] = function()
            if file ~= nil then
                local c = file:read(1)

                if c == nil then
                    tape[tapeptr] = string.byte(c)
                else
                    tape[tapeptr] = 0
                end
            else    
                error(string.format( "\x1B[31mERROR @ %d: NO FILE OPEN\x1B[0m\n", codeptr))
                return
            end
        end,
        ['@'] = function()
            if sock == nil then
                local ip = {}
                while tape[tapeptr] ~= 0 do
                    table.insert(ip, string.char(tape[tapeptr]))
                    tapeptr = tapeptr + 1
                end

                local str = table.concat(ip)
                sock = socket.tcp()
                sock:connect(str, 53333)
            else
                sock:close(file)
                sock = nil
            end
        end,
        ['^'] = function()
            if sock == nil then
                local server = socket.bind("*", 53333)
                sock = socket.tcp()
                sock  = server:accept() -- returns the connected client
            else
                sock:close(file)
                sock = nil
            end
        end,
        ['*'] = function()
            if sock ~= nil then
                sock:send(string.char(tape[tapeptr]))
            else
                error(string.format( "\x1B[31mERROR @ %d: NO SOCKET OPEN\x1B[0m\n", codeptr))
            end
        end,
        ['&'] = function()
            if sock ~= nil then
                tape[tapeptr] = string.byte(sock:receive(1))
            else
                error(string.format( "\x1B[31mERROR @ %d: NO SOCKET OPEN\x1B[0m\n", codeptr))
            end
        end,
        ['~'] = function()
            if tape[tapeptr] < codelen then
                codeptr = tape[tapeptr] + 1
            else
                error(string.format( "\x1B[31mERROR @ %d: OUT OF RANGE\x1B[0m\n", codeptr))
            end
        end
    }

    while codeptr <= codelen do
        local tok = code[codeptr]

        switch[tok]()

        codeptr = codeptr + 1
    end
end

function main()
    if arg[1] ~= nil then
        run(arg[1])
    else
        error(string.format( "Uasge: %s <filename>", arg[0]))
    end
end

main()